nav2_collision_monitor/src/range.cpp

```diff
@@ -0,0 +1,131 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/range.hpp"
+
+#include <math.h>
+#include <cmath>
+#include <functional>
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+Range::Range(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string & source_name,
+  const std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+  const std::string & base_frame_id,
+  const std::string & global_frame_id,
+  const tf2::Duration & transform_tolerance,
+  const rclcpp::Duration & source_timeout)
+: Source(
+    node, source_name, tf_buffer, base_frame_id, global_frame_id,
+    transform_tolerance, source_timeout),
+  data_(nullptr)
+{
+  RCLCPP_INFO(logger_, "[%s]: Creating Range", source_name_.c_str());
+}
+
+Range::~Range()
+{
+  RCLCPP_INFO(logger_, "[%s]: Destroying Range", source_name_.c_str());
+  data_sub_.reset();
+}
+
+void Range::configure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string source_topic;
+
+  getParameters(source_topic);
+
+  rclcpp::QoS range_qos = rclcpp::SensorDataQoS();  // set to default
+  data_sub_ = node->create_subscription<sensor_msgs::msg::Range>(
+    source_topic, range_qos,
+    std::bind(&Range::dataCallback, this, std::placeholders::_1));
+}
+
+void Range::getData(
+  const rclcpp::Time & curr_time,
+  std::vector<Point> & data) const
+{
+  // Ignore data from the source if it is not being published yet or
+  // not being published for a long time
+  if (data_ == nullptr) {
+    return;
+  }
+  if (!sourceValid(data_->header.stamp, curr_time)) {
+    return;
+  }
+
+  // Ignore data, if its range is out of scope of range sensor abilities
+  if (data_->range < data_->min_range || data_->range > data_->max_range) {
+    RCLCPP_INFO(
+      logger_,
+      "[%s]: Data range %fm is out of {%f..%f} sensor span. Ignoring...",
+      source_name_.c_str(), data_->range, data_->min_range, data_->max_range);
+    return;
+  }
+
+  // Obtaining the transform to get data from source frame and time where it was received
+  // to the base frame and current time
+  tf2::Transform tf_transform;
+  if (!getTransform(data_->header.frame_id, data_->header.stamp, curr_time, tf_transform)) {
+    return;
+  }
+
+  // Calculate poses and refill data array
+  for (
+    float angle = -data_->field_of_view / 2;
+    angle <= data_->field_of_view / 2;
+    angle += obstacles_angle_)
```

Comment:
I'm not entirely following this obstacles angle parameter, what is this exactly, and why the default that is set?

Reply:
Yes, indeed this is an issue. Thanks for noting this. [Added](https://github.com/ros-planning/navigation2/pull/3099/files#diff-8ba7a458af64954e48901b5091ebc91d61f5b592235c046e668b4e2ff82de61eR112-R123) the same code to the end of loop to check upper edge case (and corrected testcase accordingly).